client/internal/bootstrap/config_test.go

```diff
@@ -7,192 +7,181 @@ import (
-					Expect(err.Error()).To(ContainSubstring("next proto header value must be specified"))
-				})
-			})
-
-			When("AADResource is empty", func() {
-				It("should return an error", func() {
-					cfg.AADResource = ""
-					err := cfg.Validate()
-					Expect(err).ToNot(BeNil())
-					Expect(err.Error()).To(ContainSubstring("AAD resource must be specified"))
-				})
-			})
-
-			When("KubeconfigPath is empty", func() {
-				It("should return an error", func() {
-					cfg.KubeconfigPath = ""
-					err := cfg.Validate()
-					Expect(err).ToNot(BeNil())
-					Expect(err.Error()).To(ContainSubstring("kubeconfig path must be specified"))
-				})
-			})
-
-			When("CredFilePath is empty", func() {
-				It("should return an error", func() {
-					cfg.CredFilePath = ""
-					err := cfg.Validate()
-					Expect(err).ToNot(BeNil())
-					Expect(err.Error()).To(ContainSubstring("cred file path must be specified"))
-				})
-			})
-
-			When("cloud provider config path does not exist", func() {
-				It("should return an error", func() {
-					cfg.CloudProviderConfigPath = "does/not/exist.json"
-					err := cfg.Validate()
-					Expect(err).ToNot(BeNil())
-					Expect(err.Error()).To(ContainSubstring("reading cloud provider config data"))
-				})
-			})
-
-			When("cloud provider config is malformed", func() {
-				It("should return an error", func() {
-					tempDir := GinkgoT().TempDir()
-					path := filepath.Join(tempDir, "config.json")
-					err := os.WriteFile(path, []byte("malformed"), os.ModePerm)
-					Expect(err).To(BeNil())
-					cfg.CloudProviderConfigPath = path
-					err = cfg.Validate()
-					Expect(err).ToNot(BeNil())
-					Expect(err.Error()).To(ContainSubstring("unmarshalling cloud provider config data"))
-				})
-			})
-
-			When("client opts are valid", func() {
-				It("should validate without error", func() {
-					cloudProviderConfig := cloud.ProviderConfig{
-						ClientID:               "msi",
-						UserAssignedIdentityID: "identityId",
-						CloudName:              azure.PublicCloud.Name,
-					}
-					configBytes, err := json.Marshal(cloudProviderConfig)
-					Expect(err).To(BeNil())
-
-					tempDir := GinkgoT().TempDir()
-					cfg.CloudProviderConfigPath = filepath.Join(tempDir, "azure.json")
-
-					err = os.WriteFile(cfg.CloudProviderConfigPath, configBytes, os.ModePerm)
-					Expect(err).To(BeNil())
-
-					err = cfg.Validate()
-					Expect(err).To(BeNil())
-					Expect(cfg.ProviderConfig).To(Equal(cloudProviderConfig))
-				})
-			})
+				data, _ := json.Marshal(config)
+				tempDir := t.TempDir()
+				c.CloudProviderConfigPath = filepath.Join(tempDir, "azure.json")
+				_ = os.WriteFile(c.CloudProviderConfigPath, data, os.ModePerm)
+			},
+			expectedErr: "",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cfg := *baseCfg
+			tt.modify(&cfg, t)
+
+			err := cfg.Validate()
+			if tt.expectedErr != "" {
+				assert.Error(t,err)
+				assert.Contains(t, err.Error(), tt.expectedErr)
```

Comment:
there's `assert.ErrorContains` which would be more appropriate here

since we have total control of the error string you can also have expectedErr be of type `error` and simply use `assert.Equal` instead